<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# 2D Numpy in Python


Estimated time needed: **30** minutes
    

## Objectives

After completing this lab you will be able to:

* Operate comfortably with `numpy`
* Perform complex operations with `numpy`


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li><a href="#Create-a-2D-Numpy-Array">Create a 2D Numpy Array</a></li>
        <li><a href="#Accessing-different-elements-of-a-Numpy-Array">Accessing different elements of a Numpy Array</a></li>
        <li><a href="#Basic-Operations">Basic Operations</a></li>
    </ul>
  
</div>

<hr>


## Create a 2D Numpy Array


In [1]:
!pip install numpy


[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
# Import the libraries

import numpy as np

Consider the list <code>a</code>, which contains three nested lists **each of equal size**. 


In [3]:
# Create a list

a = [[11, 12, 13], [21, 22, 23], [31, 32, 33]]
a

[[11, 12, 13], [21, 22, 23], [31, 32, 33]]

We can cast the list to a Numpy Array as follows:


In [4]:
# Convert list to Numpy Array
# Every element is the same type

A = np.array(a)
A

array([[11, 12, 13],
       [21, 22, 23],
       [31, 32, 33]])

We can use the attribute <code>ndim</code> to obtain the number of axes or dimensions, referred to as the rank. 


In [5]:
# Show the numpy array dimensions

A.ndim

2

Attribute <code>shape</code> returns a tuple corresponding to the size or number of each dimension.


In [6]:
# Show the numpy array shape

A.shape

(3, 3)

The total number of elements in the array is given by the attribute <code>size</code>.


In [7]:
# Show the numpy array size

A.size

9

<hr>


## Accessing different elements of a Numpy Array


We can use rectangular brackets to access the different elements of the array. The correspondence between the rectangular brackets and the list and the rectangular representation is shown in the following figure for a 3x3 array:  


<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/NumTwoEg.png" width="500">


We can access the 2nd-row, 3rd column as shown in the following figure:


<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/NumTwoFT.png" width="400">


 We simply use the square brackets and the indices corresponding to the element we would like:


In [46]:
A[1][2]

1

In [36]:
A[1, 0]

1

 We can also use the following notation to obtain the elements: 


In [37]:
# Access the element on the second row and third column

A[1][2]

1

 Consider the elements shown in the following figure 


<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/NumTwoFF.png" width="400">


We can access the element as follows: 


In [39]:
A.shape

(2, 3)

In [42]:
A[1][2]

1

In [48]:
A[1][0]

1

In [10]:
# Access the element on the first row and first column

A[0][0]

11

We can also use slicing in numpy arrays. Consider the following figure. We would like to obtain the first two columns in the first row


<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/NumTwoFSF.png" width="400">


 This can be done with the following syntax: 


In [11]:
# Access the element on the first row and first and second columns

A[0][0:2]

array([11, 12])

In [56]:
A[1][0:2]

array([1, 0])

Similarly, we can obtain the first two rows of the 3rd column as follows:


In [12]:
# Access the element on the first and second rows and third column

A[0:2, 2]

array([13, 23])

Corresponding to the following figure: 


<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/2D_numpy.png" width="400"><br />


In [57]:
arr_2d = np.array([
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8]
])

In [58]:
arr_2d[1, :]

array([3, 4, 5])

In [65]:
arr_2d[:2, :2]

array([[0, 1],
       [3, 4]])

In [67]:
arr_2d[2:,1:2]

array([[7]])

In [68]:
arr_3d = np.array([
    [[ 0,  1,  2], [ 3,  4,  5], [ 6,  7,  8]],
    [[ 9, 10, 11], [12, 13, 14], [15, 16, 17]],
    [[18, 19, 20], [21, 22, 23], [24, 25, 26]]
])

arr_3d[1, 1, 1]

13

In [83]:
arr_3d[2:,2:, 1]

array([[25]])

In [88]:
arr_3d[1,1,2]

14

## Basic Operations


We can also add arrays. The process is identical to matrix addition. Matrix addition of <code>X</code> and <code>Y</code> is shown in the following figure:


<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/NumTwoAdd.png" width="500">


The numpy array is given by <code>X</code> and <code>Y</code>


In [92]:
# Create a numpy array X

X = np.array([[1, 0], [0, 1]]) 
X

array([[1, 0],
       [0, 1]])

In [93]:
# Create a numpy array Y

Y = np.array([[2, 1], [1, 2]]) 
Y

array([[2, 1],
       [1, 2]])

 We can add the numpy arrays as follows.


In [94]:
# Add X and Y

Z = X + Y
Z

array([[3, 1],
       [1, 3]])

Multiplying a numpy array by a scaler is identical to multiplying a matrix by a scaler. If we multiply the matrix <code>Y</code> by the scaler 2, we simply multiply every element in the matrix by 2, as shown in the figure.


<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/NumTwoDb.png" width="500">


We can perform the same operation in numpy as follows 


In [95]:
# Create a numpy array Y

Y = np.array([[2, 1], [1, 2]]) 
Y

array([[2, 1],
       [1, 2]])

In [96]:
# Multiply Y with 2

Z = 2 * Y
Z

array([[4, 2],
       [2, 4]])

Multiplication of two arrays corresponds to an element-wise product or <em>Hadamard product</em>. Consider matrix <code>X</code> and <code>Y</code>. The Hadamard product corresponds to multiplying each of the elements in the same position, i.e. multiplying elements contained in the same color boxes together. The result is a new matrix that is the same size as matrix <code>Y</code> or <code>X</code>, as shown in the following figure.


<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/NumTwoMul.png" width="500">


We can perform element-wise product of the array <code>X</code> and <code>Y</code> as follows:


In [97]:
# Create a numpy array Y

Y = np.array([[2, 1], [1, 2]]) 
Y

array([[2, 1],
       [1, 2]])

In [98]:
# Create a numpy array X

X = np.array([[1, 0], [0, 1]]) 
X

array([[1, 0],
       [0, 1]])

In [99]:
# Multiply X with Y

Z = X * Y
Z

array([[2, 0],
       [0, 2]])

We can also perform matrix multiplication with the numpy arrays <code>A</code> and <code>B</code> as follows:


First, we define matrix <code>A</code> and <code>B</code>:


In [100]:
# Create a matrix A

A = np.array([[0, 1, 1], [1, 0, 1]])
A

array([[0, 1, 1],
       [1, 0, 1]])

In [101]:
# Create a matrix B

B = np.array([[1, 1], [1, 1], [-1, 1]])
B

array([[ 1,  1],
       [ 1,  1],
       [-1,  1]])

We use the numpy function <code>dot</code> to multiply the arrays together.


In [102]:
# Calculate the dot product

Z = np.dot(A,B)
Z

array([[0, 2],
       [0, 2]])

In [103]:
# Calculate the sine of Z

np.sin(Z)

array([[0.        , 0.90929743],
       [0.        , 0.90929743]])

In [104]:
np.cos(Z)

array([[ 1.        , -0.41614684],
       [ 1.        , -0.41614684]])

In [105]:
np.tan(Z)

array([[ 0.        , -2.18503986],
       [ 0.        , -2.18503986]])

We use the numpy attribute <code>T</code> to calculate the transposed matrix


In [106]:
# Create a matrix C

C = np.array([[1,1],[2,2],[3,3]])
C

array([[1, 1],
       [2, 2],
       [3, 3]])

In [107]:
# Get the transposed of C

C.T

array([[1, 2, 3],
       [1, 2, 3]])

<h2>Quiz on 2D Numpy Array</h2>


Consider the following list <code>a</code>, convert it to Numpy Array. 


In [110]:
# Write your code below and press Shift+Enter to execute

a = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
A = np.array(a)
A

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

<details><summary>Click here for the solution</summary>

```python
A = np.array(a)
A
```

</details>


Calculate the numpy array size.


In [111]:
# Write your code below and press Shift+Enter to execute
A.size

12

<details><summary>Click here for the solution</summary>

```python
A.size
```

</details>


Access the element on the first row and first and second columns.


In [113]:
# Write your code below and press Shift+Enter to execute
A[0][0:2]

array([1, 2])

<details><summary>Click here for the solution</summary>

```python
A[0][0:2]
```

</details>


Perform matrix multiplication with the numpy arrays <code>A</code> and <code>B</code>.


In [114]:
# Write your code below and press Shift+Enter to execute

B = np.array([[0, 1], [1, 0], [1, 1], [-1, 0]])

X = np.dot(A, B)
X

array([[ 1,  4],
       [ 5, 12],
       [ 9, 20]])

<details><summary>Click here for the solution</summary>

```python
X = np.dot(A,B)
X
```

</details>


<hr>
<h2>The last exercise!</h2>
<p>Congratulations, you have completed your first lesson and hands-on lab in Python. 
<hr>


## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/" target="_blank">Joseph Santarcangelo</a>


## Other contributors

<a href="www.linkedin.com/in/jiahui-mavis-zhou-a4537814a">Mavis Zhou</a>


## Change Log


|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |
|---|---|---|---|
| 2023-11-02 | 2.2 | Abhishek Gagneja | Updated instructions |
| 2022-01-10  | 2.1  | Malika  | Removed the readme for GitShare|
| 2021-01-05 | 2.2 | Malika | Updated the solution for dot multiplication |
| 2020-09-09 | 2.1 | Malika | Updated the screenshot for first two rows of the 3rd column |
| 2020-08-26  | 2.0  | Lavanya | Moved lab to course repo in GitLab |



<hr/>

## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
